solutions/usaco-1206.mdx

```diff
@@ -0,0 +1,120 @@
+---
+id: usaco-1206
+source: USACO Silver 2022 February
+title: Redistributing Gifts
+author: Chongtian Ma
+---
+
+[Official Analysis](http://www.usaco.org/current/data/sol_prob1_silver_feb22.html)
+
+## Explanation
+
+Similar to the official analysis, we can construct a directed edge from $i$ to $j$ if cow $i$ prefers gift $j$ to gift $i$.
+If cow $i$ is able to receive a gift where it has an outgoing edge, then the solution will always be more optimal. 
+Remember to also add an edge from cow $i$ to itself in case it cannot get a better gift.
+
+Then, we can separate the graph into strongly connected components (SCCs). If $i$ is in an SCC, this means that cow $i$ can receive
+all gifts in the SCC through a series of reassignments by definition of an SCC. It is also guaranteed to receive one 
+or more preferred gifts (including the original assignment) since we are only adding these gifts as outgoing edges. If a cow does not belong in 
+an SCC, it cannot receive a more optimal reassignment.
+
+## Implementation
+
+Kosaraju's algorithm can be used to find all SCCs in the graph. At the end, it is optimal to just go down cow $i$'s preference list and assign the first gift belonging in the SCC.
```

Comment:
add a link to this website here: https://en.wikipedia.org/wiki/Kosaraju's_algorithm